tests/testthat/test_zzzz-grade_learnr.R

```diff
@@ -1,86 +1,73 @@
-  last_value <- eval(parse(text = user_code), envir = envir_result)
-
-  grade_learnr(label = "test",
-               user_code = user_code,
-               solution_code = solution_code,
-               check_code = check_code,
-               envir_result = envir_result,
-               evaluate_result = list(),
-               envir_prep = envir_prep,
-               last_value = last_value)
-}
-
 test_that("Grade learnr grade_result", {
-  expect_correct(
-    test_grade_learnr(
-      user_code = "4",
-      check_code = "grade_result(
+  expect_grade_learnr(
+    user_code = "4",
+    check_code = "grade_result(
           pass_if(~ .result == 4, 'yes. you did it.'),
           correct = 'this other correct message.',
           glue_correct = '{ .message } { .correct }'
-        )"
-    ),
-    "yes. you did it. this other correct message."
+        )",
+  is_correct = TRUE,
+  msg = "yes. you did it. this other correct message."
   )
 })
 
 test_that("Grade learnr check_code", {
-  expect_correct(
-    test_grade_learnr(
-      user_code = "4",
-      check_code = "grade_code(correct = 'This works', glue_correct = '{.correct}')",
-      solution_code = "4"
-    ),
-    "This works"
+  expect_grade_learnr(
+    user_code = "4",
+    check_code = "grade_code(correct = 'This works', glue_correct = '{.correct}')",
+    solution_code = "4",
+    is_correct = TRUE,
+    msg = "This works"
   )
 
-  expect_wrong(
-    test_grade_learnr(
-      user_code = "exp(log(2))",
-      check_code = "grade_code(glue_incorrect = '{.message}')",
-      solution_code = "exp(log(1))"
-    ),
-    "In log(2), I expected 1 where you wrote 2."
+  expect_grade_learnr(
+    user_code = "exp(log(2))",
+    check_code = "grade_code(glue_incorrect = '{.message}')",
+    solution_code = "exp(log(1))",
+    is_correct = FALSE,
+    msg = "In log(2), I expected 1 where you wrote 2."
   )
 
   # User code that produces a parsing error should return an incorrect grade (by default)
-  parse_error <- grade_learnr(user_code = "function(")
-  expect_true(is_feedback(parse_error))
-  expect_false(parse_error$correct)
-  expect_true(parse_error$type == "error")
-
+  expect_grade_learnr(
+    user_code = "function(",
+    check_code = 'function(...) stop("boom")',
+    solution_code = "4",
+    is_correct = FALSE,
+    msg = "might not be valid R code" # from default parse error in grade_learnr_parse_error()
+  )
+  
+  # Code scaffolding produces informative parsing error message
+  expect_grade_learnr(
+    user_code = "____(mtcars, cyl)",
+    is_correct = FALSE,
+    msg = "contains 1 blank"
+  )
+  
+  expect_grade_learnr(
+    user_code = "________(___, ____)",
+    is_correct = FALSE,
+    msg = "contains 3 blanks"
+  )
+  
+  # Can customize the feedback through an exercise.parse.error function
+  parse_error_func <- function(x) {
+    graded(
+      correct = FALSE,
+      message = paste(
+        "The user code of", shQuote(x$.user_code),
```

Comment:
Booo. `sQuote(q)` was add in R 3.5. 

```suggestion
        "The user code of", paste0("'", x$.user_code, "'"),
```

Reply:
Oops! Helpfully we only use it in a test 😌